File: examples/producers/typescript-minimal/src/agent.ts
/**
 * AgentLoop — TypeScript port of the Python python-minimal AgentLoop.
 *
 * Production-grade reference of the manual loop pattern with a mock LLM
 * for portability. Replace `_MockLLM` with a real client (Anthropic, OpenAI,
 * Azure OpenAI, etc.) for production use.
 */

import {
    AAEPEmitter,
    classifyErrorCategory,
    classifyRisk,
    makeId,
    safeArgsSummary,
} from "./emitter.js";
import { StreamCoalescer } from "./coalescer.js";


// === Tool descriptor ===

export interface ToolDescriptor {
    name: string;
    description: string;
    handler: (args: Record<string, unknown>) => unknown | Promise<unknown>;
    riskLevel?: "low" | "medium" | "high";
    irreversible?: boolean;
}


// === Agent loop options ===

export interface AgentLoopOptions {
    emitter: AAEPEmitter;
    tools?: ToolDescriptor[];
}


// === The agent loop ===

export class AgentLoop {
    private readonly emitter: AAEPEmitter;
    private readonly tools: Map<string, ToolDescriptor>;
    private readonly llm: _MockLLM;
    private readonly activeSessions: Map<string, AbortController> = new Map();

    constructor(options: AgentLoopOptions) {
        this.emitter = options.emitter;
        this.tools = new Map(
            (options.tools ?? this.defaultTools()).map((t) => [t.name, t]),
        );
        this.llm = new _MockLLM();
    }

    /**
     * Run a complete agent session. Returns the session_id.
     *
     * The session is fully owned by this method: it emits all lifecycle
     * events including the terminal event, even on exception.
     */
    async run(userMessage: string, opts: { userId?: string } = {}): Promise<string> {
        const sessionId = this.emitter.startSession({
            summaryNormal: `Processing: ${userMessage.slice(0, 80)}`,
            requestText: userMessage,
            requestedBy: opts.userId ? `user:${opts.userId}` : undefined,
            toolsAvailable: Array.from(this.tools.keys()),
        });

        const abortController = new AbortController();
        this.activeSessions.set(sessionId, abortController);

        try {
            await this.runLoop(sessionId, userMessage, abortController.signal);
        } finally {
            this.activeSessions.delete(sessionId);
        }

        return sessionId;
    }

    /** Cancel an in-progress session. Returns true if cancellation was issued. */
    cancel(sessionId: string): boolean {
        const controller = this.activeSessions.get(sessionId);
        if (!controller) return false;
        controller.abort();
        return true;
    }

    // === Internal: the thinking-and-acting loop ===

    private async runLoop(
        sessionId: string,
        userMessage: string,
        abortSignal: AbortSignal,
    ): Promise<void> {
        const startTime = Date.now();
        let currentState = "idle";
        let toolCount = 0;
        let coalescer: StreamCoalescer | null = null;
        const messages: Array<{ role: string; content: unknown }> = [
            { role: "user", content: userMessage },
        ];

        try {
            while (true) {
                if (abortSignal.aborted) {
                    throw new DOMException("Cancelled", "AbortError");
                }

                // Transition to thinking
                if (currentState !== "thinking") {
                    this.emitter.stateChanged({
                        sessionId,
                        fromState: currentState,
                        toState: "thinking",
                        summaryNormal: "Considering the request.",
                    });
                    currentState = "thinking";
                }

                const response = await this.llm.complete(messages, Array.from(this.tools.values()));

                if (response.toolCalls.length > 0) {
                    // Tools requested
                    this.emitter.stateChanged({
                        sessionId,
                        fromState: "thinking",
                        toState: "calling_tool",
                        summaryNormal: `Preparing to call ${response.toolCalls.length} tool(s).`,
                    });
                    currentState = "calling_tool";

                    const toolResults: Array<{ tool_call_id: string; result: unknown }> = [];
                    for (const toolCall of response.toolCalls) {
                        const result = await this.executeTool(sessionId, toolCall);
                        toolCount++;
                        toolResults.push({ tool_call_id: toolCall.id, result });
                    }

                    messages.push({
                        role: "assistant",
                        content: response.text,
                    });
                    messages.push({
                        role: "tool_results",
                        content: toolResults,
                    });
                    continue;
                }

                // No more tools; stream output
                this.emitter.stateChanged({
                    sessionId,
                    fromState: "thinking",
                    toState: "writing_output",
                    summaryNormal: "Generating response.",
                });
                currentState = "writing_output";

                const outputId = makeId("out");
                coalescer = new StreamCoalescer({
                    emitter: this.emitter,
                    sessionId,
                    outputId,
                    coalesceAt: "sentence",
                });

                for await (const chunk of this.llm.stream(response)) {
                    if (abortSignal.aborted) {
                        throw new DOMException("Cancelled", "AbortError");
                    }
                    coalescer.addToken(chunk);
                }
                coalescer.finish();
                coalescer = null;
                break;
            }

            // Success
            const durationMs = Date.now() - startTime;
            this.emitter.completeSession({
                sessionId,
                summaryNormal: "Response complete.",
                durationMs,
                toolInvocationsCount: toolCount,
            });

        } catch (error: unknown) {
            if (coalescer) {
                try {
                    coalescer.finish();
                } catch {
                    // Already finished or in bad state; swallow
                }
            }

            if (error instanceof DOMException && error.name === "AbortError") {
                this.emitter.cancelledSession({
                    sessionId,
                    cancelledBy: "system",
                    summaryNormal: "Session cancelled.",
                });
            } else {
                const err = error instanceof Error ? error : new Error(String(error));
                this.emitter.errorSession({
                    sessionId,
                    errorCategory: classifyErrorCategory(err),
                    summaryNormal: `Error: ${err.name}`,
                    errorMessage: err.message.slice(0, 1000),
                    recoverable: err.name === "TimeoutError" || err.message.toLowerCase().includes("network"),
                });
            }
            throw error;
        }
    }

    // === Internal: execute one tool with full AAEP cycle ===

    private async executeTool(
        sessionId: string,
        toolCall: { id: string; name: string; arguments: Record<string, unknown> },
    ): Promise<unknown> {
        const descriptor = this.tools.get(toolCall.name);
        const aaepCallId = makeId("call");
        const args = toolCall.arguments;

        const { riskLevel, irreversible } = descriptor
            ? {
                  riskLevel: descriptor.riskLevel ?? "low",
                  irreversible: descriptor.irreversible ?? false,
              }
            : classifyRisk(toolCall.name);

        // Emit tool.invoked BEFORE side effect
        this.emitter.toolInvoked({
            sessionId,
            tool: toolCall.name,
            toolCallId: aaepCallId,
            argsSummary: safeArgsSummary(args),
            riskLevel,
            irreversible,
            summaryNormal: `Calling ${toolCall.name}.`,
        });

        // Confirmation gating for irreversible or high-risk
        if (irreversible || riskLevel === "high") {
            const replyToken = this.emitter.awaitConfirmation({
                sessionId,
                action: `Call ${toolCall.name} with: ${safeArgsSummary(args, 200)}`,
                consequence: irreversible
                    ? "This action cannot be easily undone."
                    : "This action will be executed.",
                timeoutSeconds: 300,
                defaultDecision: "reject",
                riskLevel,
                irreversible,
                summaryNormal: `Confirm: call ${toolCall.name}?`,
            });

            const decision = await this.emitter.waitForDecision(replyToken);

            if (decision !== "accept") {
                this.emitter.toolCompleted({
                    sessionId,
                    tool: toolCall.name,
                    toolCallId: aaepCallId,
                    status: "error",
                    errorMessage: "User declined to authorize this action.",
                });
                return `<user declined to call ${toolCall.name}>`;
            }
        }

        // Execute the tool
        try {
            if (!descriptor) {
                throw new Error(`No handler for tool ${JSON.stringify(toolCall.name)}`);
            }
            const result = await descriptor.handler(args);

            this.emitter.toolCompleted({
                sessionId,
                tool: toolCall.name,
                toolCallId: aaepCallId,
                status: "success",
                summaryNormal: String(result).slice(0, 200),
            });
            return result;

        } catch (error: unknown) {
            const err = error instanceof Error ? error : new Error(String(error));
            const isTimeout = err.name === "TimeoutError" || err.message.includes("timeout");
            this.emitter.toolCompleted({
                sessionId,
                tool: toolCall.name,
                toolCallId: aaepCallId,
                status: isTimeout ? "timeout" : "error",
                errorMessage: err.message.slice(0, 1000),
            });
            throw error;
        }
    }

    // === Default mock tools ===

    private defaultTools(): ToolDescriptor[] {
        return [
            {
                name: "fetch_balance",
                description: "Look up an account balance",
                handler: async (args) => {
                    await sleep(300);
                    const account = (args.account as string) ?? "checking";
                    const balances: Record<string, string> = {
                        checking: "$3,247.18",
                        savings: "$12,891.40",
                    };
                    return balances[account] ?? "$0.00";
                },
                riskLevel: "low",
                irreversible: false,
            },
            {
                name: "send_email",
                description: "Send an email",
                handler: async (args) => {
                    await sleep(500);
                    return `Email sent to ${args.to ?? "recipient"}`;
                },
                riskLevel: "high",
                irreversible: true,
            },
            {
                name: "transfer_funds",
                description: "Move money between accounts",
                handler: async (args) => {
                    await sleep(700);
                    return `Transferred $${args.amount} from ${args.from} to ${args.to}`;
                },
                riskLevel: "high",
                irreversible: true,
            },
        ];
    }
}


// === Mock LLM ===

interface MockLLMResponse {
    text: string;
    toolCalls: Array<{ id: string; name: string; arguments: Record<string, unknown> }>;
}


class _MockLLM {
    private calls = 0;

    async complete(
        messages: Array<{ role: string; content: unknown }>,
        _tools: ToolDescriptor[],
    ): Promise<MockLLMResponse> {
        await sleep(400);
        this.calls++;

        const lastMsg = messages[messages.length - 1];
        const lastText = typeof lastMsg?.content === "string" ? lastMsg.content : "";
        const lower = lastText.toLowerCase();

        // On the first call, sometimes call a tool to exercise the flow
        if (this.calls === 1) {
            if (lower.includes("balance")) {
                return {
                    text: "Let me check that for you.",
                    toolCalls: [{
                        id: `tool_${this.calls}`,
                        name: "fetch_balance",
                        arguments: { account: "checking" },
                    }],
                };
            }
            if (lower.includes("email")) {
                return {
                    text: "I'll send that email.",
                    toolCalls: [{
                        id: `tool_${this.calls}`,
                        name: "send_email",
                        arguments: {
                            to: "recipient@example.com",
                            subject: "Re: your request",
                        },
                    }],
                };
            }
        }

        // No more tools; return final output (caller will stream the text)
        return { text: "", toolCalls: [] };
    }

    async *stream(_response: MockLLMResponse): AsyncIterable<string> {
        const chunks = [
            "Here's what I found. ",
            "Your account is in good standing ",
            "with no pending issues. ",
            "Is there anything else ",
            "I can help you with?",
        ];
        for (const chunk of chunks) {
            await sleep(50);
            yield chunk;
        }
    }
}


// === Helpers ===

function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

File: examples/producers/python-minimal/aaep_minimal_producer/agent.py
"""
Production-grade agent loop that uses AAEPEmitter to drive real sessions.

This is the manual-loop pattern from the Implementer's Guide §2.5 implemented
as a deployable reference. It includes a mock LLM client so the example runs
without depending on any real provider. Replace _MockLLM with your real client
(Anthropic, OpenAI, etc.) for production use.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Callable

from aaep_minimal_producer.emitter import (
    AAEPEmitter,
    StreamCoalescer,
    classify_error_category,
    classify_risk,
    make_id,
    safe_args_summary,
)


# === Agent loop ===

@dataclass
class ToolDescriptor:
    """Description of a tool the agent can call."""

    name: str
    description: str
    handler: Callable
    risk_level: str = "low"
    irreversible: bool = False


class AgentLoop:
    """
    A production-grade agent loop with full AAEP emission.

    This implementation is intentionally explicit. Every state transition,
    every tool call, every output emission appears in the code. If you find
    parts that should be abstracted, fork this file and abstract them — but
    the unmodified version remains the reference.
    """

    def __init__(
        self,
        emitter: AAEPEmitter,
        llm_client: Any | None = None,
        tools: list[ToolDescriptor] | None = None,
    ):
        self.emitter = emitter
        self.llm = llm_client or _MockLLM()
        self.tools: dict[str, ToolDescriptor] = {
            t.name: t for t in (tools or _default_tools())
        }
        self._active_sessions: dict[str, asyncio.Task] = {}

    async def run(
        self,
        user_message: str,
        *,
        user_id: str | None = None,
    ) -> str:
        """
        Run a complete agent session. Returns the session_id.

        The session is fully owned by this method: it emits all lifecycle
        events including the terminal event, even on exception or cancellation.
        """
        session_id = self.emitter.start_session(
            summary_normal=f"Processing: {user_message[:80]}",
            request_text=user_message,
            requested_by=f"user:{user_id}" if user_id else None,
            tools_available=list(self.tools.keys()),
        )

        # Track the session task so it can be cancelled externally
        task = asyncio.current_task()
        if task is not None:
            self._active_sessions[session_id] = task

        try:
            await self._run_loop(session_id, user_message)
        finally:
            self._active_sessions.pop(session_id, None)

        return session_id

    async def cancel(self, session_id: str) -> bool:
        """Cancel an in-progress session. Returns True if a task was cancelled."""
        task = self._active_sessions.get(session_id)
        if task is None or task.done():
            return False
        task.cancel()
        return True

    async def _run_loop(self, session_id: str, user_message: str) -> None:
        """The main thinking-and-acting loop."""
        start_time = time.monotonic()
        current_state = "idle"
        tool_count = 0
        coalescer: StreamCoalescer | None = None
        messages = [{"role": "user", "content": user_message}]

        try:
            while True:
                # Transition to thinking
                if current_state != "thinking":
                    self.emitter.state_changed(
                        session_id=session_id,
                        from_state=current_state,
                        to_state="thinking",
                        summary_normal="Considering the request.",
                    )
                    current_state = "thinking"

                response = await self.llm.complete(messages, tools=list(self.tools.values()))

                if response.tool_calls:
                    # Tools requested
                    self.emitter.state_changed(
                        session_id=session_id,
                        from_state="thinking",
                        to_state="calling_tool",
                        summary_normal=f"Preparing to call {len(response.tool_calls)} tool(s).",
                    )
                    current_state = "calling_tool"

                    tool_results = []
                    for tool_call in response.tool_calls:
                        result = await self._execute_tool(session_id, tool_call)
                        tool_count += 1
                        tool_results.append({
                            "tool_call_id": tool_call["id"],
                            "result": result,
                        })

                    messages.append({
                        "role": "assistant",
                        "content": response.text,
                        "tool_calls": response.tool_calls,
                    })
                    messages.append({"role": "tool_results", "content": tool_results})
                    continue  # back to thinking

                # No more tools; stream output
                self.emitter.state_changed(
                    session_id=session_id,
                    from_state="thinking",
                    to_state="writing_output",
                    summary_normal="Generating response.",
                )
                current_state = "writing_output"

                output_id = make_id("out")
                coalescer = StreamCoalescer(
                    emitter=self.emitter,
                    session_id=session_id,
                    output_id=output_id,
                    coalesce_at="sentence",
                )

                async for chunk in self.llm.stream(response):
                    coalescer.add_token(chunk)
                coalescer.finish()
                coalescer = None
                break

            # Success
            duration_ms = int((time.monotonic() - start_time) * 1000)
            self.emitter.complete_session(
                session_id=session_id,
                summary_normal="Response complete.",
                duration_ms=duration_ms,
                tool_invocations_count=tool_count,
            )

        except asyncio.CancelledError:
            if coalescer is not None:
                coalescer.finish()
            self.emitter.cancelled_session(
                session_id=session_id,
                cancelled_by="system",
                summary_normal="Session cancelled.",
            )
            raise

        except Exception as exc:
            if coalescer is not None:
                try:
                    coalescer.finish()
                except Exception:
                    pass
            self.emitter.error_session(
                session_id=session_id,
                error_category=classify_error_category(exc),
                summary_normal=f"Error: {type(exc).__name__}",
                error_message=str(exc)[:1000],
                recoverable=isinstance(exc, (TimeoutError, ConnectionError)),
            )
            raise

    async def _execute_tool(self, session_id: str, tool_call: dict[str, Any]) -> Any:
        """Execute one tool with full AAEP emission cycle."""
        tool_name = tool_call["name"]
        descriptor = self.tools.get(tool_name)
        aaep_tool_call_id = make_id("call")
        args = tool_call.get("arguments", {})

        # Use descriptor metadata when available; fall back to name-based heuristic
        if descriptor:
            risk_level = descriptor.risk_level
            irreversible = descriptor.irreversible
        else:
            risk_level, irreversible = classify_risk(tool_name)

        # Emit tool.invoked BEFORE any side effect
        self.emitter.tool_invoked(
            session_id=session_id,
            tool=tool_name,
            tool_call_id=aaep_tool_call_id,
            args_summary=safe_args_summary(args),
            risk_level=risk_level,
            irreversible=irreversible,
            summary_normal=f"Calling {tool_name}.",
        )

        # Confirmation gating for irreversible or high-risk
        if irreversible or risk_level == "high":
            reply_token = self.emitter.await_confirmation(
                session_id=session_id,
                action=f"Call {tool_name} with arguments: {safe_args_summary(args, 200)}",
                consequence=(
                    "This action cannot be easily undone."
                    if irreversible
                    else "This action will be executed."
                ),
                timeout_seconds=300,
                default_decision="reject",
                risk_level=risk_level,
                irreversible=irreversible,
                summary_normal=f"Confirm: call {tool_name}?",
            )

            try:
                decision = await asyncio.wait_for(
                    self.emitter.wait_for_decision(reply_token),
                    timeout=300,
                )
            except asyncio.TimeoutError:
                decision = "reject"  # timeout applies default_decision

            if decision != "accept":
                self.emitter.tool_completed(
                    session_id=session_id,
                    tool=tool_name,
                    tool_call_id=aaep_tool_call_id,
                    status="error",
                    error_message="User declined to authorize this action.",
                )
                return f"<user declined to call {tool_name}>"

        # Execute the tool
        try:
            if descriptor:
                result = await _maybe_await(descriptor.handler(**args))
            else:
                result = f"<no handler for tool {tool_name!r}>"

            self.emitter.tool_completed(
                session_id=session_id,
                tool=tool_name,
                tool_call_id=aaep_tool_call_id,
                status="success",
                summary_normal=str(result)[:200],
            )
            return result

        except asyncio.TimeoutError:
            self.emitter.tool_completed(
                session_id=session_id,
                tool=tool_name,
                tool_call_id=aaep_tool_call_id,
                status="timeout",
                error_message="Tool exceeded timeout",
            )
            raise

        except Exception as exc:
            self.emitter.tool_completed(
                session_id=session_id,
                tool=tool_name,
                tool_call_id=aaep_tool_call_id,
                status="error",
                error_message=str(exc)[:1000],
            )
            raise


# === Helpers ===

async def _maybe_await(value: Any) -> Any:
    """Await value if it's awaitable, otherwise return as-is."""
    if asyncio.iscoroutine(value):
        return await value
    return value


def _default_tools() -> list[ToolDescriptor]:
    """Mock tool registry used by the example. Replace for real usage."""
    return [
        ToolDescriptor(
            name="fetch_balance",
            description="Look up an account balance",
            handler=_mock_fetch_balance,
            risk_level="low",
            irreversible=False,
        ),
        ToolDescriptor(
            name="search_records",
            description="Search internal records",
            handler=_mock_search_records,
            risk_level="low",
            irreversible=False,
        ),
        ToolDescriptor(
            name="send_email",
            description="Send an email",
            handler=_mock_send_email,
            risk_level="high",
            irreversible=True,
        ),
        ToolDescriptor(
            name="transfer_funds",
            description="Move money between accounts",
            handler=_mock_transfer_funds,
            risk_level="high",
            irreversible=True,
        ),
    ]


async def _mock_fetch_balance(account: str = "checking") -> str:
    await asyncio.sleep(0.3)
    balances = {"checking": "$3,247.18", "savings": "$12,891.40"}
    return balances.get(account, "$0.00")


async def _mock_search_records(query: str = "") -> str:
    await asyncio.sleep(0.5)
    return f"Found 3 records matching '{query}'"


async def _mock_send_email(to: str = "", subject: str = "", body: str = "") -> str:
    await asyncio.sleep(0.5)
    return f"Email sent to {to} with subject '{subject}'"


async def _mock_transfer_funds(
    from_account: str = "", to_account: str = "", amount: float = 0.0
) -> str:
    await asyncio.sleep(0.7)
    return f"Transferred ${amount:.2f} from {from_account} to {to_account}"


# === Mock LLM ===

@dataclass
class _LLMResponse:
    """A mock LLM response."""

    text: str = ""
    tool_calls: list[dict[str, Any]] = field(default_factory=list)


class _MockLLM:
    """
    A minimal mock LLM client used by the example. Not a real model;
    just produces realistic-looking agent behavior to exercise AAEP emissions.

    Replace with a real client (Anthropic, OpenAI, etc.) for production.
    """

    def __init__(self) -> None:
        self._calls = 0

    async def complete(self, messages: list[dict], tools: list = None) -> _LLMResponse:
        """Pretend to think; sometimes call a tool, sometimes finish."""
        await asyncio.sleep(0.4)
        self._calls += 1

        # On the first call, sometimes call a tool to demonstrate the flow
        last_message = messages[-1].get("content", "") if messages else ""
        wants_tool = (
            self._calls == 1
            and isinstance(last_message, str)
            and any(kw in last_message.lower() for kw in (
                "balance", "search", "email", "transfer", "send", "find"
            ))
        )

        if wants_tool:
            # Pick a relevant mock tool to call
            if "balance" in last_message.lower():
                return _LLMResponse(
                    text="Let me check that for you.",
                    tool_calls=[{
                        "id": f"tool_{self._calls}",
                        "name": "fetch_balance",
                        "arguments": {"account": "checking"},
                    }],
                )
            if "email" in last_message.lower():
                return _LLMResponse(
                    text="I'll send that email.",
                    tool_calls=[{
                        "id": f"tool_{self._calls}",
                        "name": "send_email",
                        "arguments": {
                            "to": "recipient@example.com",
                            "subject": "Re: your request",
                            "body": "Following up as requested.",
                        },
                    }],
                )
            if "transfer" in last_message.lower():
                return _LLMResponse(
                    text="I'll prepare that transfer.",
                    tool_calls=[{
                        "id": f"tool_{self._calls}",
                        "name": "transfer_funds",
                        "arguments": {
                            "from_account": "checking-7821",
                            "to_account": "savings-3344",
                            "amount": 500.00,
                        },
                    }],
                )

        # Final output
        return _LLMResponse(text="", tool_calls=[])

    async def stream(self, response: _LLMResponse) -> AsyncIterator[str]:
        """Stream the response text token by token."""
        chunks = (
            "Here's what I found. ",
            "Your account is in good standing ",
            "with no pending issues. ",
            "Is there anything else ",
            "I can help you with?",
        )
        for chunk in chunks:
            await asyncio.sleep(0.05)
            yield chunk
